### PR TITLE
fix(NODE-4413): set maxTimeMS on getMores when maxAwaitTimeMS is specified

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

- Capture maxAwaitTimeMS option on cursor options
- Add documentation to options
- Move error handling to getMore.execute (I can put this back, it seems like the right place in terms of handling issues)
- GetMore constructor must always take in options
- Added tests to maxTimeMS.test.ts and converted to TS
- Added tests to change streams.

#### What is the motivation for this change?

Fix maxAwaitTimeMS not being respected by change stream getMores (all getMores were impacted actually)

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/integration/crud/maxTimeMS.test.ts [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    let cursor: FindCursor<{ _id: number }>;
+    let events: CommandStartedEvent[];
+
+    beforeEach(async function () {
+      client = this.configuration.newClient({ monitorCommands: true });
+      await client
+        .db()
+        .dropCollection('cappedAt3')
+        .catch(() => null);
+      await sleep(100); // Need to make sure it has had time to drop entirely
```
Comment (baileympearson) [inline test/integration/crud/maxTimeMS.test.ts:144]: Why?  Does dropCollection return eagerly?

File: test/integration/crud/maxTimeMS.test.ts [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    afterEach(async function () {
+      events = [];
+      await cursor?.close();
+      await client?.close();
+    });
+
+    const tailableValues = [true, false, undefined];
+    const awaitDataValues = [true, false, undefined];
+    const maxTimeMSValues = [100, 0, undefined];
+    const maxAwaitTimeMSValues = [100, 0, undefined];
```
Comment (baileympearson) [inline test/integration/crud/maxTimeMS.test.ts:165]: How were these values chosen?  Is it a >0, 0 and undefined value?  

Should we also be testing with negative values?

File: src/cursor/abstract_cursor.ts [hunk 3]
```diff
@@ -617,21 +645,8 @@ export abstract class AbstractCursor<
...
-    }
-
-    if (server == null) {
-      callback(new MongoRuntimeError('Unable to iterate cursor without selected server'));
-      return;
-    }
-
-    const getMoreOperation = new GetMoreOperation(cursorNs, cursorId, server, {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const getMoreOperation = new GetMoreOperation(this[kNamespace], this[kId]!, this[kServer]!, {
```
Comment (baileympearson) [inline src/cursor/abstract_cursor.ts:649]: Do we have tests that test for errors being thrown when execute is called on a zero-ed cursor or if the server is undefined?  If not, can we add them to demonstrate this behavior still exists?

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/integration/crud/maxTimeMS.test.ts (hunk 1 above)
Comment (nbbeeken) [reply test/integration/crud/maxTimeMS.test.ts]: I shouldn't need this, in sharded cluster topologies reusing a namespace that was just dropped isn't reccomended IIRC and I thought I was running into that. dropped.

File: test/integration/crud/maxTimeMS.test.ts (hunk 2 above)
Comment (nbbeeken) [reply test/integration/crud/maxTimeMS.test.ts:164]: I don't think we need negative tests here, MongoDB will error. I tested zero because its a falsey case, and 100 is just a postive value

File: src/cursor/abstract_cursor.ts (hunk 3 above)
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:649]: Added! :)

File: test/unit/operations/get_more.test.ts [hunk 4]
```diff
@@ -192,6 +192,54 @@ describe('GetMoreOperation', function () {
         }
       });
     });
+
+    context('error cases', () => {
```
Comment (dariakp) [inline test/unit/operations/get_more.test.ts:196]: what about if server is undefined, or was that code always unreachable?

File: test/integration/crud/maxTimeMS.test.ts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  MongoServerError
+} from '../../../src';
+import { assert as test, setupDatabase } from '../shared';
+
+describe('MaxTimeMS', function () {
+  before(function () {
+    return setupDatabase(this.configuration);
+  });
+
+  it('should correctly respect the maxTimeMS property on count', function (done) {
```
Comment (dariakp) [inline test/integration/crud/maxTimeMS.test.ts:19]: can we do async/await in these tests? moving them to ts already effectively makes it a brand new file to review and it's only a handful of tests so shouldn't add much scope to refactor but will make it easier to review the actual tests

**Review by dariakp — CHANGES_REQUESTED** [review]

File: test/unit/operations/get_more.test.ts (hunk 4 above)
Comment (nbbeeken) [reply test/unit/operations/get_more.test.ts:196]: Should always be unreachable, server selection would have to fail, and we would have something wrong in our error handling of server selection if we proceeded to execute with a nullish server.

File: test/integration/crud/maxTimeMS.test.ts (hunk 5 above)
Comment (nbbeeken) [reply test/integration/crud/maxTimeMS.test.ts]: Done!

File: test/integration/change-streams/change_stream.test.ts [hunk 6]
```diff
@@ -1372,23 +1372,98 @@ describe('Change Streams', function () {
...
+        .expectEvents({
+          client: 'client0',
+          ignoreExtraEvents: true, // Sharded clusters have extra getMores
+          events: [
+            { commandStartedEvent: { commandName: 'aggregate', command: { maxTimeMS: 5000 } } },
+            { commandStartedEvent: { commandName: 'insert' } },
+            {
+              commandStartedEvent: {
+                commandName: 'getMore',
+                command: { maxTimeMS: { $$exists: false } }
```
Comment (dariakp) [inline test/integration/change-streams/change_stream.test.ts:1442]: it's not clear from the test case name why it should not exist on the aggregate command but not the getMore command here

File: test/integration/change-streams/change_stream.test.ts [hunk 7]
```diff
@@ -1372,23 +1372,98 @@ describe('Change Streams', function () {
...
+        })
+        .operation({
+          object: 'changeStreamOnClient',
+          name: 'iterateUntilDocumentOrError',
+          ignoreResultAndError: true
+        })
+        .expectEvents({
+          client: 'client0',
+          events: [
+            { commandStartedEvent: { commandName: 'aggregate' } },
```
Comment (dariakp) [inline test/integration/change-streams/change_stream.test.ts:1407]: why do we assert absence of the prop on the getMore in the 2nd case, but not on the initial aggregate here? what's the actual expected behavior from these options in the two cases?

File: test/integration/change-streams/change_stream.test.ts (hunk 6 above)
Comment (nbbeeken) [reply test/integration/change-streams/change_stream.test.ts:1451]: Added a case to demonstrate what should happen when both options are provided, as well as updated test names.

File: test/integration/change-streams/change_stream.test.ts (hunk 7 above)
Comment (nbbeeken) [reply test/integration/change-streams/change_stream.test.ts]: Added an (not) existence check, and clarified the titles

**Review by baileympearson — APPROVED** [review]

**Review by dariakp — APPROVED** [review]
